protocols/request-response/src/codec/header.rs

```diff
@@ -0,0 +1,244 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use async_trait::async_trait;
+use bytes::{Bytes, BytesMut};
+use futures::prelude::*;
+use libp2p_core::ProtocolName;
+use minicbor::{Encode, Decode};
+use std::io;
+use super::RequestResponseCodec;
+use unsigned_varint::{aio, io::ReadError};
+
+/// A protocol header.
+#[derive(Debug, Default, Clone, PartialEq, Eq, Encode, Decode)]
+#[cbor(map)]
+pub struct Header {
+    /// The type of message.
+    #[n(0)] pub typ: Option<Type>,
+    /// The number of additional requests the remote is willing to receive.
+    #[n(1)] pub credit: Option<u16>,
+    /// An identifier used for sending credit grants.
+    #[n(2)] pub ident: Option<u64>
+}
+
+/// A protocol message type.
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
+pub enum Type {
+    #[n(0)] Request,
+    #[n(1)] Response,
+    #[n(2)] Credit,
+    #[n(3)] Ack
+}
+
+/// A protocol message consisting of header and data.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Message<T> {
+    header: Header,
+    data: Option<T>
+}
+
+impl<T> Message<T> {
+    /// Create a new message of some type.
+    fn new(header: Header) -> Self {
+        Message { header, data: None }
+    }
+
+    /// Create a request message.
+    pub fn request(data: T) -> Self {
+        let mut m = Message::new(Header { typ: Some(Type::Request), .. Header::default() });
+        m.data = Some(data);
+        m
+    }
+
+    /// Create a response message.
+    pub fn response(data: T) -> Self {
+        let mut m = Message::new(Header { typ: Some(Type::Response), .. Header::default() });
+        m.data = Some(data);
+        m
+    }
+
+    /// Create a credit grant.
+    pub fn credit(credit: u16, ident: u64) -> Self {
+        Message::new(Header { typ: Some(Type::Credit), credit: Some(credit), ident: Some(ident) })
+    }
+
+    /// Create an acknowledge message.
+    pub fn ack(ident: u64) -> Self {
+        Message::new(Header { typ: Some(Type::Ack), credit: None, ident: Some(ident) })
+    }
+
+    /// Access the message header.
+    pub fn header(&self) -> &Header {
+        &self.header
+    }
+
+    /// Access the message data.
+    pub fn data(&self) -> Option<&T> {
+        self.data.as_ref()
+    }
+
+    /// Consume this message and return header and data.
+    pub fn into_parts(self) -> (Header, Option<T>) {
+        (self.header, self.data)
+    }
+}
+
+/// A wrapper around a `ProtocolName` impl which augments the protocol name.
+///
+/// The type implements `ProtocolName` itself and creates a name for a
+/// request-response protocol based on the protocol name of the wrapped type.
+#[derive(Debug, Clone)]
+pub struct ProtocolWrapper<P>(P, Bytes);
+
+impl<P: ProtocolName> ProtocolWrapper<P> {
+    pub fn v1(p: P) -> Self {
+        let mut full = BytesMut::from(b"/request-response/header/1".as_ref());
```

Comment:
```suggestion
        let mut full = BytesMut::from(b"/request-response/throttled/1".as_ref());
```

Reply:
I wonder if we should really use such long prefixes which are negotiated all the time. Since the actual protocol can be assumed to be fairly unique, I switched to a short `/t/1` prefix instead. Currently the multistream-select related data is sometimes larger than the actual request-response data.

Reply:
Yes, I was thinking the same, though my first instinct was to just drop (or shorten to `/rr` or `/rqrs`) the `/request-response` prefix, also because in principle if there are multiple such wrapper protocols they may even be stacked on top of each other, which would repeat the `request-response` prefix (e.g. `/throttled/1/foobar/1` rather than `/request-response/throttled/1/request-response/foobar/1`).